transport/src/test/java/io/netty/channel/AbstractChannelTest.java

```diff
@@ -82,6 +86,69 @@ public void ensureDefaultChannelId() {
         assertTrue(channelId instanceof DefaultChannelId);
     }
 
+    @Test
+    public void testClosedChannelExceptionCarryIOException() throws Exception {
+        final IOException ioException = new IOException();
+        final Channel channel = new TestChannel() {
+            private boolean open = true;
+            private boolean active;
+
+            @Override
+            protected AbstractUnsafe newUnsafe() {
+                return new AbstractUnsafe() {
+                    @Override
+                    public void connect(
+                            SocketAddress remoteAddress, SocketAddress localAddress, ChannelPromise promise) {
+                        active = true;
+                        promise.setSuccess();
+                    }
+                };
+            }
+
+            @Override
+            protected void doClose()  {
+                active = false;
+                open = false;
+            }
+
+            @Override
+            protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+                throw ioException;
+            }
+
+            @Override
+            public boolean isOpen() {
+                return open;
+            }
+
+            @Override
+            public boolean isActive() {
+                return active;
+            }
+        };
+
+        EventLoop loop = new DefaultEventLoop();
+        try {
+            registerChannel(loop, channel);
+            channel.connect(new InetSocketAddress(NetUtil.LOCALHOST, 8888)).sync();
```

Comment:
this will spuriously succeed on some machines won't it?

Reply:
It does not matter what we pass here as I implement `AbstractUnsafe.connect(...)` which will just notify the promise. Remember this is no socket etc.